jupyter nemo tutorial (for issue #848)

rupertford opened the pull request.

rupertford commented:
@arporter. This is ready for review. As it is a tutorial I suggest you simply make any changes you think will benefit it as part of the review rather than passing it back to me, unless there are serious changes required.

arporter commented on tutorial/nemo/nemo_example1.ipynb:
```diff
@@ -0,0 +1,316 @@
+   "metadata": {},
+   "source": [
+    "Also notice that the outer loops have indices `jk` with bounds from `1` to `jpk`, the middle loops have indices `jj` with bounds from `1` to `jpj` and the inner loops have indices `ji` with bounds from `1` to `jpi`. The NEMO coding rules enforce this naming convention throughout the code where `ji` corresponds to iterating over longitude, `jj` corresponds to iterating over latitude and `jk`  ...
+   ]
```
PSYCLONE_CONFIG is not normally set. Are you OK if I replace this with a link to the appropriate part of the UG?

rupertford replied:
Ah. Yes sure. I just thought they might like to have a look for the file and see the content.

rupertford replied:
Yes sure. I just thought they might want to see the actual file.

arporter replied:
I've put a link to the file in the github repo - finding where the "actual file" is installed is tricky so I thought this would be simpler.

rupertford replied:
Yes that's a good solution

arporter commented:
I've added a section with a link to Binder to the top-level README.md. It works(!) but will need updating to point to master rather than this branch when this is merged.
I've added a couple of sections on PSyIR navigation but now realise that I've not added a link to them from within the NEMO tutorial - I'll do that in a mo. If you could take a look at those sections that would be great.

We should probably add automatic validation of these tutorials to Travis (in the same way that the notebooks in the examples are run). I'm not sure what the nicest way of doing this is - I might be able to extend the functionality under `examples` to go back up to the `tutorial` directory and do its magic there...probably requires a few Makefiles to be copy-n-pasted.

Finally, it occurred to me that @deardenchris would be an ideal reviewer for this as he has no/little existing knowledge of PSyclone :-) (But probably best to wait until Rupert and I have convinced ourselves that it's ready...)

deardenchris commented:
Hi @arporter yes happy to review this - let me know when you're ready for me to do so

rupertford commented:
> If you could take a look at those sections that would be great.

They look really good. My only comments are 1) you should perhaps tell them that they've finished this section at the end, 2) when you give them to option to do something themselves it's probably worth having an empty code box beneath, 3) I think the brackets in "(nested loops)" should be "(nested) loops", 4) do you think we should explain "canonical" and perhaps "language-agnostic", or am I being too low-brow, 5) perhaps refer back to the nemo tutorial when mentioning the use of NEMO in this section? and 6) we should probably be consistent in how we suggest moving on to the next example.

arporter commented:
> 4) do you think we should explain "canonical" and perhaps "language-agnostic", or am I being too low-brow, 

Oh yes, I was intending to do that but then that opens up a can of notebooks in that I'd want to explain what happens to `else if` and `where` statements. Actually, I'll write some words and basically point them to the User Guide/Developer Guide.

arporter commented:
Thanks for those suggestions - I've done all of them. If you're happy @rupertford, would you mind requesting a review from Chris?

rupertford commented:
This is ready for review. As agreed @deardenchris would be a good person to review this and he has kindly accepted. The original idea was for people taking the tutorial to have a local copy and run jupyter locally (`jupyter-notebook`). However @arporter has also added a link to the README which allows it to be run directly from git using binder. I think both approaches need to be tested :-). Note, the binder link in the README would need to be updated before merging to master.

rupertford commented:
@deardenchris, I've given you write access to the repo. That should allow you to review (and eventually merge), but if not, let me know.

deardenchris commented:
Hi @rupertford thanks, I can see there is now the option for me to add my review. I'll be sure to test the tutorial both locally and using Binder. I've done code reviews before but this will be my first using GitHub so I might come back to you or @arporter before finalising just to make sure I'm following the correct procedures :-)

arporter commented:
> 
> 
> Hi @rupertford thanks, I can see there is now the option for me to add my review. I'll be sure to test the tutorial both locally and using Binder. I've done code reviews before but this will be my first using GitHub so I might come back to you or @arporter before finalising just to make sure I'm following the correct procedures :-)

Don't worry too much about 'correct procedures' for this - as it's a tutorial rather than code we can be a lot more relaxed. Really, we just need to make sure that what we've written makes sense to someone new to PSyclone. Please add comments in-line (if that works OK for notebook files) or just collate and paste them here.

deardenchris commented:
Hi @rupertford and @arporter I've now gone through the tutorial notebooks and I think they are really good. The code examples are nice and simple to understand, and the accompanying text does a good job of explaining what each bit is doing.

I had a slight issue running the notebooks locally, which I think is a macOS-specific problem. I installed PSyclone locally on my MacBook using 'pip3 install .' , and then initiated the tutorial thus (from top-level dir of PSyclone repo):

jupyter notebook tutorial/introduction.ipynb

Everything was fine until I reached the 1st NEMO example, 3rd code box, which creates the PSyIR and associated objects. I got the error 'PSyclone configuration error: psyclone.cfg not found'. After some investigation, it seems that the psyclone.cfg file was not in any of the default paths, but instead had been placed within /usr/local/share/psyclone/. As a workaround, I simply copied the file from this location to ~/.local/share/psyclone, and this allowed me to complete the rest of the tutorial. 

I also tested the Binder versions and they work faultlessly. I like the colour-coding of the PSyIR output in the Binder notebooks as it makes it even easier to read. 

A few very minor things I noticed: 

1) In NEMO example 3, there is a typo in the second text block (scedule --> schedule)
2) In the fparser2 notebook, it might be worth adding in an empty code cell where it is suggested that the users have a go themselves at printing out the DO statements
3) Similarly in NEMO example 3, add an empty code cell where it is suggested that users attempt to parallelise over levels
4) In README.md, add the 'tutorial' sub-directory to the Structure section at the bottom
5) As already mentioned, the binder link in the README will need to be updated before merging onto master

rupertford commented:
Ready for next review from @deardenchris subject to travis being happy. I've hopefully addressed the issues raised. I've also added a small section in the user guide and updated setup.py to include the tutorial in the distribution. The update of setup.py probably needs @arporter to check over so I humbly suggest passing it on to him once you are happy with the tutorial part.

rupertford commented:
It works for me with a local installation ...

```
rupert@ubuntu:~/proj/PSyclone$ ls ~/.local/share/psyclone
ls: cannot access '/home/rupert/.local/share/psyclone': No such file or directory
rupert@ubuntu:~/proj/PSyclone$ pip install . --user
Processing /home/rupert/proj/PSyclone
Collecting pyparsing (from PSyclone==1.9.0)
  Using cached https://files.pythonhosted.org/packages/8a/bb/488841f56197b13700afd5658fc279a2025a39e22449b7cf29864669b15d/pyparsing-2.4.7-py2.py3-none-any.whl
Collecting fparser==0.0.11 (from PSyclone==1.9.0)
Collecting configparser (from PSyclone==1.9.0)
  Using cached https://files.pythonhosted.org/packages/7a/2a/95ed0501cf5d8709490b1d3a3f9b5cf340da6c433f896bbe9ce08dbe6785/configparser-4.0.2-py2.py3-none-any.whl
Collecting six (from PSyclone==1.9.0)
  Using cached https://files.pythonhosted.org/packages/ee/ff/48bde5c0f013094d729fe4b0316ba2a24774b3ff1c52d924a8a4cb04078a/six-1.15.0-py2.py3-none-any.whl
Collecting enum34 (from PSyclone==1.9.0)
  Using cached https://files.pythonhosted.org/packages/6f/2c/a9386903ece2ea85e9807e0e062174dc26fdce8b05f216d00491be29fad5/enum34-1.1.10-py2-none-any.whl
Installing collected packages: pyparsing, six, fparser, configparser, enum34, PSyclone
  Running setup.py install for PSyclone ... done
Successfully installed PSyclone-1.9.0 configparser-4.0.2 enum34-1.1.10 fparser-0.0.11 pyparsing-2.4.7 six-1.15.0
rupert@ubuntu:~/proj/PSyclone$ ls ~/.local/share/psyclone
examples  psyclone.cfg  tutorial
rupert@ubuntu:~/proj/PSyclone$ 
```

rupertford commented:
venv works too ...

```
rupert@ubuntu:~/proj/PSyclone$ export WORKON_HOME=~/Envs
rupert@ubuntu:~/proj/PSyclone$ workon py3
(py3) rupert@ubuntu:~/proj/PSyclone$ pip uninstall psyclone
Uninstalling PSyclone-1.9.0:
  Would remove:
    /home/rupert/Envs/py3/lib/python3.6/site-packages/PSyclone.egg-link
Proceed (y/n)? y
  Successfully uninstalled PSyclone-1.9.0
(py3) rupert@ubuntu:~/proj/PSyclone$ ls /home/rupert/Envs/py3/share/psyclone
ls: cannot access '/home/rupert/Envs/py3/share/psyclone': No such file or directory
(py3) rupert@ubuntu:~/proj/PSyclone$ pip install .
Processing /home/rupert/proj/PSyclone
Requirement already satisfied: pyparsing in /home/rupert/Envs/py3/lib/python3.6/site-packages (from PSyclone==1.9.0) (2.4.0)
Requirement already satisfied: fparser==0.0.11 in /home/rupert/Envs/py3/lib/python3.6/site-packages (from PSyclone==1.9.0) (0.0.11)
Requirement already satisfied: configparser in /home/rupert/Envs/py3/lib/python3.6/site-packages (from PSyclone==1.9.0) (3.7.4)
Requirement already satisfied: six in /home/rupert/Envs/py3/lib/python3.6/site-packages (from PSyclone==1.9.0) (1.12.0)
Building wheels for collected packages: PSyclone
  Building wheel for PSyclone (setup.py) ... done
  Stored in directory: /tmp/pip-ephem-wheel-cache-ujr4mvjq/wheels/6b/d7/7b/5ed541068f663d7671056820e1dc18a73af0f29122a58010f9
Successfully built PSyclone
Installing collected packages: PSyclone
Successfully installed PSyclone-1.9.0
(py3) rupert@ubuntu:~/proj/PSyclone$ ls /home/rupert/Envs/py3/share/psyclone
examples  psyclone.cfg  tutorial
(py3) rupert@ubuntu:~/proj/PSyclone$ 
```

deardenchris commented:
Thanks for checking Rupert. I can't spot anything obvious that I've done differently myself. I have an Ubuntu VM installed on my macbook so I will try to install on there... Will let you know what happens

deardenchris commented:
For reference this is what I get on my Mac. Only thing that looks different is the python version and the versions of the dependencies...

`chrisdearden@deardenhost:~/git_repos/PSyclone$ pip3 uninstall psyclone`
`WARNING: Skipping psyclone as it is not installed.`
`chrisdearden@deardenhost:~/git_repos/PSyclone$ pip3 install .`
`Processing /Users/chrisdearden/git_repos/PSyclone`
`Requirement already satisfied: pyparsing in /Users/chrisdearden/Library/Python/3.7/lib/python/site-packages (from PSyclone==1.9.0) (2.4.7)`
`Requirement already satisfied: fparser==0.0.11 in /Users/chrisdearden/Library/Python/3.7/lib/python/site-packages (from PSyclone==1.9.0) (0.0.11)`
`Requirement already satisfied: configparser in /Users/chrisdearden/Library/Python/3.7/lib/python/site-packages (from PSyclone==1.9.0) (5.0.0)`
`Requirement already satisfied: six in /Users/chrisdearden/Library/Python/3.7/lib/python/site-packages (from PSyclone==1.9.0) (1.15.0)`
`Building wheels for collected packages: PSyclone`
 `Building wheel for PSyclone (setup.py) ... done`
  `Created wheel for PSyclone: filename=PSyclone-1.9.0-cp37-none-any.whl size=521164 sha256=fc1efb0a9d48bea4015d10901087cb5c7184191d6bc399abbf92228cd3dcc48d
  Stored in directory: /private/var/folders/cc/ts0__ln96019z_brslqngky80000gq/T/pip-ephem-wheel-cache-ov15rft6/wheels/e6/30/ff/b02f8c49889d1e8c711f8ee3d63bb7ede0cb95fe7076c32e71`
`Successfully built PSyclone`
`Installing collected packages: PSyclone`
`Successfully installed PSyclone-1.9.0`
`chrisdearden@deardenhost:~/git_repos/PSyclone$ ls -lrt /usr/local/share/psyclone/`
`total 8.0K`
`-rw-r--r-- 1 chrisdearden 7.0K Jul 23 11:22 psyclone.cfg`

rupertford commented:
You're not getting the examples either!

deardenchris commented:
Yes very strange. I can confirm that it works fine though on my Ubuntu 20.4 virtual machine, with both the `examples` and `tutorial` sub-directories visible within ~/.local/share/psyclone after installation.

deardenchris commented:
@rupertford @arporter Managed to get this working on my Mac by switching to using `python setup.py install` instead of pip. Not exactly sure why the pip method isn't working properly for me though, could be something weird going on with my local setup

rupertford commented:
@deardenchris has determined that the problem with the MAC pre-dates this PR and is going to add a separate issue. Therefore I'm putting this back to ready for review for @arporter to take a look.

deardenchris commented:
New issue describing the pip problem now raised separately as https://github.com/stfc/PSyclone/issues/861
